src/main.rs

```diff
@@ -394,7 +408,61 @@ async fn run() -> Result<()> {
         PathBuf::from(&src)
     };
 
-    let mut choices = Vec::new();
+    // Let's first find all subdirectories that are compatible with our supported recipes.
+    let root_src_dir = src_dir.clone();
+    let mut open_subdirs = vec![(0, src_dir)];
+    let mut sourceroot_candidates = vec![];
+    while !open_subdirs.is_empty() {
+        // If we call this, we have a non-empty vector.
+        let (depth, target_dir) = open_subdirs.pop().unwrap();
+        let subdirs = std::fs::read_dir(&target_dir)
+            .context(format!("failed to list subdirectories of {}", target_dir.to_string_lossy()))?;
+
+        for maybe_subdir in subdirs {
+            let subdir = maybe_subdir.context(
+                format!("unexpected io error while reading a subdirectory of {}", target_dir.to_string_lossy())
+            )?;
+
+            // We want only subdirectories.
+            if !subdir.path().is_dir() {
```

Comment:
`subdir.path()` can be extracted into a variable here since it is used multiple times, and it seems to actually do something
https://github.com/rust-lang/rust/blob/a0111af531178e48375f14f838d7a2298524067c/library/std/src/sys/unix/fs.rs#L792-L794